paasta_tools/contrib/check_orphans.py

```diff
@@ -0,0 +1,213 @@
+    responses = {}
+    for idx in range(0, len(tasks), CHUNK_SIZE):
+        resp = await asyncio.gather(*tasks[idx:idx + CHUNK_SIZE], return_exceptions=True)
+        responses.update(dict(resp))
+    return responses
+
+
+def read_one_nerve_file(nerve_config: str) -> Set[InstanceTuple]:
+    services = json.loads(nerve_config)['services']
+    return {
+        InstanceTuple(service['host'], service['port'], service['zk_path'][len(PREFIX):])
+        for service in services.values()
+        if service['zk_path'].startswith(PREFIX)
+    }
+
+
+def read_nerve_files(nerve_configs: Dict[str, Optional[str]]) -> Tuple[Set[InstanceTuple], Set[str]]:
+    instance_set: Set[InstanceTuple] = set()
+    not_found_hosts: Set[str] = set()
+    for host, host_config in nerve_configs.items():
+        if host_config is None:
+            not_found_hosts.add(host)
+        else:
+            instance_set |= read_one_nerve_file(host_config)
+    return instance_set, not_found_hosts
+
+
+def get_instance_data(service_blacklist_path: str) -> Tuple[Set[InstanceTuple], Set[InstanceTuple]]:
+    blacklisted_services = set(yaml.load(service_blacklist_path))
+    # Dump ZK
+    zk_data = get_zk_data(blacklisted_services)
+    zk_instance_data = read_from_zk_data(zk_data)
+
+    hosts = {x[0] for x in zk_instance_data}
+
+    # Dump Nerve configs from each host via xinetd
+    coro = gather_files(hosts)
+    results = asyncio.get_event_loop().run_until_complete(coro)
+
+    nerve_instance_data, not_found_hosts = read_nerve_files(results)
+
+    # Filter out anything that we couldn't get a nerve config for
+    zk_instance_data_filtered = {x for x in zk_instance_data if x[0] not in not_found_hosts}
+
+    logging.info('zk_instance_data (unfiltered) len: {}'.format(len(zk_instance_data)))
+    logging.info('zk_instance_data (filtered) len: {}'.format(len(zk_instance_data_filtered)))
+    logging.info('nerve_instance_data len: {}'.format(len(nerve_instance_data)))
+
+    return zk_instance_data_filtered, nerve_instance_data
+
+
+def write_output(
+    output_dir: str,
+    zk_instance_data: Set[InstanceTuple],
+    nerve_instance_data: Set[InstanceTuple],
+) -> None:
+    orphans = zk_instance_data - nerve_instance_data
+    if orphans:
+        logging.warning('{} orphans found'.format(len(orphans)))
+
+    pathlib.Path(output_dir).mkdir(exist_ok=True)
+
+    # groupby host
+
+    orphans_by_host: DefaultDict[str, List[Tuple[int, str]]] = defaultdict(list)
+    for orphan in orphans:
+        orphans_by_host[orphan.host].append((orphan.port, orphan.service))
+    with open(os.path.join(output_dir, 'orphans_by_host.yaml'), 'w') as f:
+        yaml.dump(dict(orphans_by_host), f)
+
+    # groupby service
+    orphans_by_service: DefaultDict[str, List[Tuple[str, int]]] = defaultdict(list)
+    for orphan in orphans:
+        orphans_by_service[orphan.service].append((orphan.host, orphan.port))
+    with open(os.path.join(output_dir, 'orphans_by_service.yaml'), 'w') as f:
```

Comment:
why do we need a yaml file?, here and line 187?

Reply:
This was originally for my own benefit, but I think is good for whatever checker script to output the orphans. To address your previous comment as well, I wrote this so that it isn't meant to act as a check itself, rather it would be used by a check.

Reply:
Why though?

Even still, you could just run the command and `>` the output and make the check look for a string in the output?

But I say just run the thing directly.

Reply:
I strongly dislike doing string matching on stdout as a signal. I tend not to prefer the unix-y exit-code-as-a-signal approach, but I think I feel much less strongly about it as you do, so i'll go ahead and change it to have the exit code signal the existence of the orphans

Reply:
Thank you